java-checks/src/main/java/org/sonar/java/checks/DeadStoreCheck.java

```diff
@@ -0,0 +1,228 @@
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.cfg.LiveVariables;
+import org.sonar.java.cfg.LocalVariableReadExtractor;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.java.resolve.SemanticModel;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Rule(
+  key = "S1854",
+  name = "Dead stores should be removed",
+  tags = {"cert", "cwe", "suspicious", "unused"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.DATA_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class DeadStoreCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (methodTree.block() == null) {
+      return;
+    }
+
+    // TODO(npe) Exclude try statements with finally as CFG is incorrect for those and lead to false positive
+    if (hasTryFinallyWithLocalVar(methodTree.block(), methodTree.symbol())) {
+      return;
+    }
+
+    Symbol.MethodSymbol methodSymbol = methodTree.symbol();
+    CFG cfg = CFG.build(methodTree);
+    LiveVariables liveVariables = LiveVariables.analyze(cfg);
+    // Liveness analysis provides information only for block boundaries, so we should do analysis between elements within blocks
+    for (CFG.Block block : cfg.blocks()) {
+      checkElements(block, liveVariables.getOut(block), methodSymbol);
+    }
+  }
+
+  private void checkElements(CFG.Block block, Set<Symbol> blockOut, Symbol.MethodSymbol methodSymbol) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "checkElements" is 30 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)